contracts/contracts/token/OUSD.sol

```diff
@@ -584,20 +594,100 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
             ? MAX_SUPPLY
             : _newTotalSupply;
 
-        _rebasingCreditsPerToken = _rebasingCredits.divPrecisely(
-            _totalSupply.sub(nonRebasingSupply)
-        );
+        _rebasingCreditsPerToken =
+            (_rebasingCredits * 1e18) /
+            (_totalSupply - nonRebasingSupply);
 
         require(_rebasingCreditsPerToken > 0, "Invalid change in supply");
 
-        _totalSupply = _rebasingCredits
-            .divPrecisely(_rebasingCreditsPerToken)
-            .add(nonRebasingSupply);
+        _totalSupply =
+            ((_rebasingCredits * 1e18) / _rebasingCreditsPerToken) +
+            nonRebasingSupply;
 
         emit TotalSupplyUpdatedHighres(
             _totalSupply,
             _rebasingCredits,
             _rebasingCreditsPerToken
         );
     }
+
+    function delegateYield(address from, address to) external onlyGovernor {
+        require(from != address(0), "Zero from address not allowed");
+        require(to != address(0), "Zero to address not allowed");
+
+        require(from != to, "Cannot delegate to self");
+        require(
+            yieldFrom[to] == address(0) &&
+                yieldTo[to] == address(0) &&
+                yieldFrom[from] == address(0) &&
+                yieldTo[from] == address(0),
+            "Blocked by existing yield delegation"
+        );
+        RebaseOptions stateFrom = rebaseState[from];
+        RebaseOptions stateTo = rebaseState[to];
+
+        require(
+            stateFrom == RebaseOptions.NotSet ||
+                stateFrom == RebaseOptions.StdNonRebasing ||
+                stateFrom == RebaseOptions.StdRebasing,
+            "Invalid rebaseState from"
+        );
+
+        require(
+            stateTo == RebaseOptions.NotSet ||
+                stateTo == RebaseOptions.StdNonRebasing ||
+                stateTo == RebaseOptions.StdRebasing,
+            "Invalid rebaseState to"
+        );
+
+        if (alternativeCreditsPerToken[from] == 0) {
+            _rebaseOptOut(from);
+        }
+        if (alternativeCreditsPerToken[to] > 0) {
+            _rebaseOptIn(to);
+        }
+
+        // Set up the bidirectional links
+        yieldTo[from] = to;
+        yieldFrom[to] = from;
+        rebaseState[from] = RebaseOptions.YieldDelegationSource;
+        rebaseState[to] = RebaseOptions.YieldDelegationTarget;
+
+        uint256 balance = balanceOf(from);
```

Comment:
🔴 This scares me badly - I'm impressed we haven't found a bug from it in testing, but we are first partially altering the accounting on both accounts, then doing a balanceOf one in an intermediate state. This is how you get bad bugs. We should move the `balance = ` and `credits = ` up to above the first write on the bidirectional links.

Reply:
good catch. There were no bugs because account transitioning to yield delegation source doesn't change the account's balance. Especially now where with your changes where there are no rounding errors. I agree it feels a lot less scary having these executed before the rebase state changes and after the accounts have their rabaseOptIn / rebaseOptOut changed (when necessary): https://github.com/OriginProtocol/origin-dollar/pull/2298/commits/f8280d13e2f5411f1da0c636f22f333e1cbcce06